feat(activity): realize the activity functionality

luckygalaxy666 opened the pull request:
## Try to implement activity mode like GitHub

**Finished:**

**Database SQL**
- `t_activity`
- `t_activity_assign_label`
- `t_activity_comment`
- `t_activity_designate_assignee`
- `t_label`

**Repository Label API**
- `REPOSITORY_CREATE_LABEL_API_PATH` = `/api/repository/create-label`
- `REPOSITORY_UPDATE_LABEL_API_PATH` = `/api/repository/update-label`
- `REPOSITORY_DELETE_LABEL_API_PATH` = `/api/repository/delete-label`
- `REPOSITORY_PAGE_LABEL_API_PATH` = `/api/repository/page-label`

**Activity API**
- `ACTIVITY_GET_ACTIVITY_API_PATH` = `/api/activity/get`
- `ACTIVITY_CREATE_ACTIVITY_API_PATH` = `/api/activity/create`
- `ACTIVITY_DELETE_ACTIVITY_API_PATH` = `/api/activity/delete`
- `ACTIVITY_UPDATE_ACTIVITY_CONTENT_API_PATH` = `/api/activity/update-content`
- `ACTIVITY_UPDATE_ACTIVITY_LOCKED_STATE_API_PATH` = `/api/activity/update-lock-state`
- `ACTIVITY_UPDATE_ACTIVITY_CLOSED_STATE_API_PATH` = `/api/activity/update-close-state`
- `ACTIVITY_PAGE_ACTIVITY_API_PATH` = `/api/activity/page`

**Activity Label API**
- `ACTIVITY_ADD_LABEL_API_PATH` = `/api/activity/add-label`
- `ACTIVITY_REMOVE_LABEL_API_PATH` = `/api/activity/delete-label`
- `ACTIVITY_PAGE_LABEL_API_PATH` = `/api/activity/page-label`

**Activity Comment API**
- `ACTIVITY_CREATE_COMMENT_API_PATH` = `/api/activity/create-comment`
- `ACTIVITY_UPDATE_COMMENT_CONTENT_API_PATH` = `/api/activity/update-comment-content`
- `ACTIVITY_UPDATE_COMMENT_RESOLVED_STATE_API_PATH` = `/api/activity/update-comment-resolved-state`
- `ACTIVITY_UPDATE_COMMENT_HIDDEN_STATE_API_PATH` = `/api/activity/update-comment-hidden-state`
- `ACTIVITY_DELETE_COMMENT_API_PATH` = `/api/activity/delete-comment`
- `ACTIVITY_PAGE_COMMENT_API_PATH` = `/api/activity/page-comment`
- `ACTIVITY_PAGE_SUB_COMMENT_API_PATH` = `/api/activity/page-sub-comment`

**Activity Assignee API**
- `ACTIVITY_ADD_ASSIGNEE_API_PATH` = `/api/activity/add-assignee`
- `ACTIVITY_REMOVE_ASSIGNEE_API_PATH` = `/api/activity/delete-assignee`
- `ACTIVITY_PAGE_ASSIGNEE_API_PATH` = `/api/activity/page-assignee`

**Sub Issue API**
- `ACTIVITY_PAGE_SUB_ISSUE_API_PATH` = `/api/activity/page-sub-issue`
- `ACTIVITY_CREATE_SUB_ISSUE_API_PATH` = `/api/activity/create-sub-issue`
- `ACTIVITY_ADD_SUB_ISSUE_API_PATH` = `/api/activity/add-sub-issue`
- `ACTIVITY_REMOVE_SUB_ISSUE_API_PATH` = `/api/activity/remove-sub-issue`

**Resolve issue:** #153

**Further:**
- Draft Activity
- Pull Request Module
- User API for Activity (if needed)

**Example: Activity Detail Page**

![Activity Detail 1](https://github.com/user-attachments/assets/30ea49ff-f67f-427f-978a-eaefbc017bbb)  
![Activity Detail 2](https://github.com/user-attachments/assets/02d80820-dfb5-4961-9426-aa1336a51fa5)  
![Activity Detail 3](https://github.com/user-attachments/assets/719eca34-c95f-4c1f-bcd4-5e061ae15885)

luckygalaxy666 committed:
feat(activity): realize the activity functionality

Try to implement activity mode like Github.

Finished:
* Database SQL
* Repository Label API
* Activity API
* Activity Label API
* Activity Comment API
* Activity Assignee API
* Some Repository-related tables information is not deleted when deleting
a repository #153

Further:
* Sub Activity
* Draft Activity
* User API for Activity(if need)

Kaiser-Yang commented on database/init/001_t_activity_label.sql:
```diff
@@ -0,0 +1,17 @@
+    id bigint NOT NULL,
+    activity_id bigint NOT NULL,
+    label_id smallint NOT NULL,
+    gmt_created timestamp without time zone DEFAULT CURRENT_TIMESTAMP NOT NULL,
```
增加 `user_id` 表示添加者。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/constant/ApiPathConstant.java:
```diff
@@ -64,4 +64,13 @@ public class ApiPathConstant {
+  public static final String ISSUE_UPDATE_ISSUE_API_PATH = ISSUE_API_PREFIX + "/update";
+  public static final String ISSUE_PAGE_ISSUE_API_PATH = ISSUE_API_PREFIX + "/page";
+  public static final String ISSUE_CHECK_ISSUE_TITLE_VALIDITY_API_PATH =
+      ISSUE_API_PREFIX + "/issue-title";
```
这个接口怎么检查？检查哪这方面？

luckygalaxy666 replied:
删除`/page`，复制多了

Kaiser-Yang replied:
你删除/page干什么，我标的是/issue-title

luckygalaxy666 replied:
已删除`issue-title`

Kaiser-Yang commented:
你用了 `label_id`，但是没有 `label` 表。

Kaiser-Yang commented:
你的提交信息格式也不太对。

Kaiser-Yang commented on database/init/001_t_repository_activity.sql:
```diff
@@ -0,0 +1,35 @@
+    activity_type smallint NOT NULL, -- 0: issue, 1: pull_request
+    issue_type_id smallint, -- only in issue
+    is_closed boolean DEFAULT false,
+    is_conversation_locked boolean DEFAULT false,
```
使用 gmt_locked 来表示是否锁定。

luckygalaxy666 commented:
> 你用了 `label_id`，但是没有 `label` 表。

添加`label`表

Kaiser-Yang commented:
你应该在push后再点resolve conversation。

luckygalaxy666 commented:
哈哈，收到，打扰了
&nbsp;


393628658
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Qingzhou ***@***.***&gt;; 
发送时间: 2025年5月22日(星期四) 下午5:14
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [CMIPT/gcs-back-end] Issues Mode   (PR #149)












Kaiser-Yang left a comment (CMIPT/gcs-back-end#149)
 
你应该在push后再点resolve conversation。
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.Message ID: ***@***.***&gt;

Kaiser-Yang commented on database/init/001_t_activity_assignee.sql:
```diff
@@ -0,0 +1,18 @@
+CREATE TABLE public.t_activity_assignee (
+    id bigint NOT NULL,
+    activity_id bigint NOT NULL,
+    user_id smallint NOT NULL,
```
use bigint here

Kaiser-Yang commented on database/init/001_t_activity_comment.sql:
```diff
@@ -0,0 +1,21 @@
+    id bigint NOT NULL,
+    activity_id bigint NOT NULL,
+    user_id bigint NOT NULL,
+    content text ,
```
NOT NULL here.

Kaiser-Yang commented on database/init/001_t_activity_label.sql:
```diff
@@ -0,0 +1,19 @@
+CREATE TABLE public.t_activity_label (
```
Rename this table.

Kaiser-Yang commented on database/init/001_t_issue_relationship.sql:
```diff
@@ -0,0 +1,18 @@
+CREATE TABLE public.t_issue_relationship (
```
Remove this table.

Kaiser-Yang commented on database/init/001_t_label.sql:
```diff
@@ -0,0 +1,21 @@
+CREATE TABLE public.t_label (
```
Use t_activity_label here

Kaiser-Yang commented on database/init/001_t_label.sql:
```diff
@@ -0,0 +1,19 @@
+     id bigint NOT NULL,
+     label_name character varying(255) NOT NULL,
+     label_description character varying(255) NOT NULL,
+     label_color character varying(255) NOT NULL,
```
update with label_hex_color, and check how long a hex color is.

Kaiser-Yang replied:
Add user_id to specify the creator, and repository_id.

luckygalaxy666 committed:
format(all): apply google java format automatically

Kaiser-Yang commented on database/init/002_all_column_constraint.sql:
```diff
@@ -32,3 +32,43 @@ ALTER TABLE ONLY public.t_user_collaborate_repository
 ALTER TABLE ONLY public.t_user_collaborate_repository
```
将constraint移动到对应的table中。

luckygalaxy666 replied:
已解决

Kaiser-Yang commented on database/init/001_t_activity.sql:
```diff
@@ -0,0 +1,35 @@
+COMMENT ON COLUMN public.t_activity.gmt_locked IS 'Timestamp when the activity was locked. NULL if not locked.';
+COMMENT ON COLUMN public.t_activity.gmt_deleted IS 'Timestamp when the activity was deleted. NULL if not deleted.';
+
+
```
移除多余换行

luckygalaxy666 committed:
format(database): move constraints to the corresponding table

By the way, solve a little bug of labelHexColor and labelDescription
misalignment.

luckygalaxy666 committed:
format(all): apply google java format automatically

Kaiser-Yang commented on database/init/001_t_activity.sql:
```diff
@@ -0,0 +1,41 @@
+ALTER TABLE ONLY public.t_activity
+    ADD CONSTRAINT t_activity_number_repository_id
+    UNIQUE (number, repository_id, gmt_deleted);
+CREATE INDEX idx_repo_activity ON public.t_activity(repository_id, number DESC);
```
为什么要在这里创建一个索引？

Kaiser-Yang commented on database/init/001_t_label.sql:
```diff
@@ -0,0 +1,36 @@
+    UNIQUE (name, repository_id, gmt_deleted);
+ALTER TABLE ONLY public.t_label
+    ADD CONSTRAINT unique_t_label_color_repository_id
+    UNIQUE (hex_color, repository_id, gmt_deleted);
```
颜色应该是可以重复的。

Kaiser-Yang commented on database/init/001_t_ssh_key.sql:
```diff
@@ -1,19 +1,29 @@
-COMMENT ON COLUMN public.t_ssh_key.gmt_updated IS 'Timestamp when the ssh_key record was last updated.';
-COMMENT ON COLUMN public.t_ssh_key.gmt_deleted IS 'Timestamp when the ssh_key record was deleted.
-If set to NULL, it indicates that the ssh_key record has not been deleted.';
+CREATE TABLE public.t_ssh_key (
```
不应该修改原始内容。

Kaiser-Yang commented on database/init/001_t_user_collaborate_repository.sql:
```diff
@@ -1,18 +1,25 @@
-COMMENT ON COLUMN public.t_user_collaborate_repository.gmt_updated IS 'Timestamp when the relationship was last updated.';
-COMMENT ON COLUMN public.t_user_collaborate_repository.gmt_deleted IS 'Timestamp when the relationship was deleted.
-If set to NULL, it indicates that the repository has not been deleted.';
+CREATE TABLE public.t_user_collaborate_repository (
```
不应该修改原始内容。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/controller/ActivityController.java:
```diff
@@ -0,0 +1,645 @@
+            == null) {
+      logger.debug(
+          "User[{}] tried to get repository of user[{}]", userId, repositoryPO.getUserId());
+      if (repositoryPO.getIsPrivate()) throw e;
```
用大括号括起来。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/controller/ActivityController.java:
```diff
@@ -0,0 +1,645 @@
+   *
+   * @param repositoryId the repository id
+   * @param userId the user id
+   * @param e the message when the repository is not found
```
你注释和方法也对不上啊

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/controller/ActivityController.java:
```diff
@@ -0,0 +1,645 @@
+    if (repositoryPO == null) {
+      throw new GenericException(ErrorCodeEnum.REPOSITORY_NOT_FOUND, repositoryId);
+    }
+    // If the repository is private, we return NOT_FOUND to make sure the user can't know
```
注释和实现对不上。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/controller/ActivityController.java:
```diff
@@ -0,0 +1,645 @@
+    }
+  }
+
+  private void checkActivityOperationValidity(
```
当仓库是公有的，活动的创建者应该是可以修改活动的。
当仓库是私有的，只是仓库的创建都和合作者可以修改活动。

luckygalaxy666 commented on src/main/java/edu/cmipt/gcs/controller/ActivityController.java:
```diff
@@ -0,0 +1,645 @@
+  @Autowired private ActivityDesignateAssigneeService activityDesignateAssigneeService;
+  @Autowired private LabelService labelService;
+
+  // TODO 查询问题子问题
```
还没做，后续实现

Kaiser-Yang replied:
后续是指之后的pr还是过两天，如何是之后的pr的话，这个可以在你的pr的描述里面记录一下之后实现的功能。

luckygalaxy666 commented on src/main/java/edu/cmipt/gcs/controller/ActivityController.java:
```diff
@@ -0,0 +1,645 @@
+        description = "Activity delete failed",
+        content = @Content(schema = @Schema(implementation = ErrorVO.class)))
+  })
+  public void deleteActivity(
```
![image](https://github.com/user-attachments/assets/dc05aa7c-5e77-4187-928c-e5558becbd7b)
似乎应该按照这个介绍写，只有仓库创建者才可以删除活动，这样的话就不需要检查可见性。

luckygalaxy666 commented on src/main/java/edu/cmipt/gcs/controller/ActivityController.java:
```diff
@@ -0,0 +1,645 @@
+    } else {
+      activityNumber = latestActivityPO.getNumber() + 1;
+    }
+    var activityPO = new ActivityPO(activity, activityNumber, idInToken.toString());
```
![image](https://github.com/user-attachments/assets/c769a28c-65b7-4e57-972a-d8b5b47d2d9d)
是有唯一索引的，但是我发现查询最新`number`时由于`mybatis-plus`的逻辑删除优化，无法查询到已删除的活动的`number`,所以要修改为原生sql；此外，因为唯一索引的缘故，并发时只会有一个活动创建成功，我会在下次提交中加入`number+1`重试策略，尽量并发时都会成功。

luckygalaxy666 committed:
fix(database&&activity): modify content based on comment requests

In this commit:
* move constraint to corresbonding sql table.
* remove `getActivityComment` API.
* move `checkActivityOperationValidity` and
  `checkRepositoryOperationValidity` into `ActivityServive` to reuse.
* make the activity creator can modify activity now.
* realize `delete acitivity`, but throw exception.
* fix some code formatting.

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/enumeration/ActivityOrderByEnum.java:
```diff
@@ -0,0 +1,7 @@
+package edu.cmipt.gcs.enumeration;
+
+public enum ActivityOrderByEnum {
```
活动是不是应该增加按照最新的评论时间排序的功能？

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/pojo/label/ActivityAssignLabelPO.java:
```diff
@@ -0,0 +1,35 @@
+
+  @TableLogic private Timestamp gmtDeleted;
+
+  public ActivityAssignLabelPO(Long userId, Long activityId, Long labelId) {
```
这里是不是可以直接调委托构建？

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/pojo/assign/ActivityDesignateAssigneePO.java:
```diff
@@ -0,0 +1,35 @@
+
+  @TableLogic private Timestamp gmtDeleted;
+
+  public ActivityDesignateAssigneePO(Long idInToken, Long activityId, Long assigneeId) {
```
这里是不是可以用委托构建？

Kaiser-Yang commented:
在之前和新增的PO类里面，如果出现了没有调用委托构建函数的情况，应该更改为调用委托构造，例如：

```java
// 不能提取的变量先用null代替
this(null, xxx, xxx.. );
// 再进行相关的提取操作
this.id = 123L;
```

luckygalaxy666 committed:
fix(activity): use delegating constructors for PO objects

Besides,fix some bugs raised by kaiser-yang.

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/pojo/activity/ActivityPO.java:
```diff
@@ -0,0 +1,63 @@
+
+  @TableLogic private Timestamp gmtDeleted;
+
+  public ActivityPO(ActivityDTO activityDTO, Integer activityNumber, String userId) {
```
使用委托构建。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/pojo/comment/CommentDTO.java:
```diff
@@ -0,0 +1,41 @@
+        @NotBlank(groups = {CreateGroup.class, UpdateGroup.class})
+        String content,
+
+    // TODO 检验？
```
检验可以在controller里面做，或者你可以学习一下自定义validator

luckygalaxy666 replied:
这个功能下个pr实现吧，我现在还不太会。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/pojo/activity/ActivityDetailVO.java:
```diff
@@ -0,0 +1,37 @@
+    @Schema(description = "Activity Description") String description,
+    @Schema(description = "Activity Labels") List<LabelVO> labels,
+    @Schema(description = "Username") String username,
+    @Schema(description = "Everyone's name of Assignees") List<AssigneeVO> assignees,
```
直接用Assignees就行，你这里是个AssigneeVO出不是String嘛。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/pojo/activity/ActivityPO.java:
```diff
@@ -0,0 +1,71 @@
+        null,
+        null,
+        null);
+    try {
```
这种代码好像出现的比较多，你可以封闭成一个函数放在一个工具类里面例如：
```java
Long convertToLong(String value, bool throwWhenFailed =  false, Long default = null);
```

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/pojo/activity/ActivityQueryDTO.java:
```diff
@@ -0,0 +1,35 @@
+
+@Schema(description = "Activity Page Query Data Transfer Object")
+public record ActivityQueryDTO(
+    @Schema(description = "User Information", example = "Name/ID/Token")
```
这里的example应该给出具体值的例子。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/pojo/activity/ActivityQueryDTO.java:
```diff
@@ -0,0 +1,35 @@
+        String user,
+    @Schema(description = "Type of User", example = "name") @NotNull(groups = QueryGroup.class)
+        UserQueryTypeEnum userType,
+    @Schema(description = "Repository ID", example = "11") @NotNull(groups = QueryGroup.class)
```
你可以不给这个例子，雪花算法不可能有id为11的情况。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/pojo/activity/ActivityQueryDTO.java:
```diff
@@ -0,0 +1,35 @@
+        UserQueryTypeEnum userType,
+    @Schema(description = "Repository ID", example = "11") @NotNull(groups = QueryGroup.class)
+        String repositoryId,
+    @Schema(description = "Author of the Activity", example = "author123") String author,
```
如果 这里是表示的用户名，你在description里面应该写成：The name of the author

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/pojo/activity/ActivityQueryDTO.java:
```diff
@@ -0,0 +1,35 @@
+    @Schema(description = "Repository ID", example = "11") @NotNull(groups = QueryGroup.class)
+        String repositoryId,
+    @Schema(description = "Author of the Activity", example = "author123") String author,
+    @Schema(description = "is the Activity a Pull Request?", example = "true")
```
Whether or not the activity is a pull request

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/pojo/activity/ActivityQueryDTO.java:
```diff
@@ -0,0 +1,35 @@
+            description = "Labels associated with the Activity",
+            example = "[\"bug\", \"enhancement\"]")
+        List<String> labels,
+    @Schema(description = "Assignees of the Activity", example = "[\"assignee1\", \"assignee2\"]")
```
如果名字应该在description里面明确写出Assignees' names of the activity

luckygalaxy666 commented on src/main/java/edu/cmipt/gcs/service/PermissionServiceImpl.java:
```diff
@@ -0,0 +1,60 @@
+      logger.debug(
+          "User[{}] tried to get repository of user[{}]", userId, repositoryPO.getUserId());
+      // If the repository is private, we return NOT_FOUND to make sure the user can't know
+      if (repositoryPO.getIsPrivate()) {
```
因为传进来的异常一定是xxx_NOT_FOUND, 包括但不限于`COMMENT_NOT_FOUND`,`REPOSITORY_NOT_FOUND`,`ACTIVITY_NOT_FOUND`
![image](https://github.com/user-attachments/assets/4b808e2f-b088-4af3-aa08-6a243c279cf9)
![image](https://github.com/user-attachments/assets/dec3ee72-f8d7-4acb-9f18-73a240a9610b)

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/service/PermissionService.java:
```diff
@@ -0,0 +1,22 @@
+   * @param repositoryId the repository id
+   * @param userId the user id
+   * @param operationTypeEnum the operation type
+   * @param e the message when the repository is not found
```
这也应该抛repository not found。

luckygalaxy666 committed:
refactor(activity):refactor some schema descriptions and code

In this commit, we refactor some schema descriptions and add
`TypeConversionUtil` to realize String -> Long

luckygalaxy666 committed:
format(all): apply google java format automatically

luckygalaxy666 committed:
refactor(repository): delete unused constructor

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/pojo/comment/CommentPO.java:
```diff
@@ -0,0 +1,74 @@
+    this.activityId = TypeConversionUtil.convertToLong(comment.activityId());
+    this.parentId = TypeConversionUtil.convertToLong(comment.parentId());
+    this.gmtResolved = getResolvedTimeSinceEpoch(comment.isResolved());
+    this.gmtHidden = getHiddenTimeSinceEpoch(comment.isHidden());
```
这几个获取时间的方法应该放到comment的类里面，我上会说的也是放在comment里面吧。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/pojo/repository/RepositoryPO.java:
```diff
@@ -37,34 +38,33 @@ public class RepositoryPO {
+        null);
+    this.id = TypeConversionUtil.convertToLong(repositoryDTO.id());
+    this.userId = TypeConversionUtil.convertToLong(userId);
     if (this.repositoryDescription == null) {
```
这里如果数据库添加默认值就可以删除这两个if的判断。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/service/PermissionServiceImpl.java:
```diff
@@ -0,0 +1,60 @@
+  public void checkActivityOperationValidity(
+      Long activityId, Long idInToken, OperationTypeEnum operationTypeEnum, GenericException e) {
+    var activityPO = activityService.getById(activityId);
+    if (activityPO == null) {
```
所以你应该限制前端同时把仓库信息传上。

luckygalaxy666 replied:
虽然我仍然认为我目前的实现是没有问题的，不过毫无疑问加上仓库id会让逻辑更清晰，所以在之后的提交我会修改所有活动方法的请求参数。

luckygalaxy666 replied:
经过讨论，因为存在仓库->活动->标签这样的逻辑链，如果每一级都要带上前面所有级的相关信息，对于方法所需的参数将会非常多，目前还是不改变现有方法的实现。如果后续有更好的方法或者必须要求前端携带仓库信息时再修改逻辑。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/service/LabelService.java:
```diff
@@ -0,0 +1,12 @@
+public interface LabelService extends IService<LabelPO> {
+  LabelPO getOneByNameAndRepositoryId(String labelName, Long repositoryId);
+
+  LabelPO getOneByHexColorAndRepositoryId(String labelHexColor, Long repositoryId);
```
hexcolor可以重复，这里为什么是getone

luckygalaxy666 commented on src/main/java/edu/cmipt/gcs/service/ActivityServiceImpl.java:
```diff
@@ -0,0 +1,216 @@
+    Page<ActivityDetailDTO> activityDetailDTOPage =
+        activityMapper.selectJoinPage(page, ActivityDetailDTO.class, queryWrapper);
+    logger.info("total activities: {}", activityDetailDTOPage.getTotal());
+    logger.info("record activities: {}", activityDetailDTOPage.getRecords());
```
这里是我自己测试效果时，忘记删了

luckygalaxy666 commented on src/main/java/edu/cmipt/gcs/service/ActivityServiceImpl.java:
```diff
@@ -0,0 +1,216 @@
+  }
+
+  @Override
+  public Page<ActivityDetailDTO> pageActivities(
```
之前用过ai提供的方法，没有好的结果，所以暂时先采用这种方法吧。如果后续我找到了更好的方法再修改，先把现有其他错误的东西先改正做说吧。

luckygalaxy666 commented on src/main/java/edu/cmipt/gcs/service/ActivityServiceImpl.java:
```diff
@@ -0,0 +1,216 @@
+            .toList();
+    if (!activityIds.isEmpty()) {
+      // 删除活动相关的信息（标签、指定的参与者、评论等）
+      activityAssignLabelService.removeByActivityIds(activityIds);
```
因为活动没有标签导致的一条没删也会返回false,而这种情况是允许的，所以我没有检查返回值

luckygalaxy666 commented on src/main/java/edu/cmipt/gcs/pojo/comment/CommentVO.java:
```diff
@@ -0,0 +1,30 @@
+    @Schema(description = "Name of the file where the comment is made. NULL if not applicable")
+        String codePath,
+    @Schema(description = "Selected line from code") Integer codeLine,
+    @Schema(description = "Parent Comment ID") String parentId,
```
这里的parentId，可以替换为`inReplyTo`,表明是某个评论的回复。
![image](https://github.com/user-attachments/assets/95d2c851-95af-4b49-9d68-ae81920f5805)
目前确实在父评论删除时会出现问题。
我研究了一下github的实现方法，在对评论进行回复时，将回复的`inReplyTo`设置为该评论段中的第一个评论（根评论）。如果根评论被删除，则会评论失败如图中所示。
<img width="747" alt="1751779617382" src="https://github.com/user-attachments/assets/0643480f-9c04-4036-bd1c-ce5cf54fabc6" />
所以在每次删除评论时，要额外判断是否是根评论，如果是根评论则要查找所有子评论回复，并将回复中按照创建时间的第一个作为新的根评论，并更新其他回复的`inReplyTO`,然后将新的根评论id返回给前端。

luckygalaxy666 committed:
fix(activity): fix some problems mentioned in comments

In this commit:
* add default value of repository_description,so that we don't need to add
  value in repositoryPO manually.
* make hexcolor can repeatable and remove some check method.
* use String.valueOf instead of Objects.toString.
* add update Activity and Comment state API

luckygalaxy666 committed:
refactor(all): use TypeconversionUtil.convertToLong

In this commit:
* Replace all methods involving Long.valueOf() with
  TypeconversionUtil.convertToLong().
* Build the page object inside the service method.
* fix some minor problems mentions in comments

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/controller/ActivityController.java:
```diff
@@ -0,0 +1,544 @@
+        description = "User activities page failed",
+        content = @Content(schema = @Schema(implementation = ErrorVO.class)))
+  })
+  public PageVO<ActivityDetailVO> pageActivity(
```
page拿到的应该不是detail，detail一般是指我点进一个评论后的内容。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/controller/ActivityController.java:
```diff
@@ -0,0 +1,544 @@
+      @Validated(QueryGroup.class) @RequestBody ActivityQueryDTO activityQueryDTO,
+      @RequestHeader(HeaderParameter.ACCESS_TOKEN) String accessToken) {
+    Long repositoryId = null;
+    try {
```
可以使用工具类。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/controller/ActivityController.java:
```diff
@@ -0,0 +1,544 @@
+      logger.error(e.getMessage());
+      throw new GenericException(ErrorCodeEnum.MESSAGE_CONVERSION_ERROR);
+    }
+    String user = activityQueryDTO.user();
```
为什么查一个仓库的活动需要用到user的信息？token里面不是有user的信息吗？

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/controller/ActivityController.java:
```diff
@@ -0,0 +1,544 @@
+  public ActivityDetailVO getActivityDetail(
+      @RequestParam(value = "id", required = false) Long id,
+      @RequestParam(value = "activityNumber", required = false) Long activityNumber,
+      @RequestParam(value = "repositoryId", required = false) Long repositoryId,
```
这样的话可能这一类方法单独封闭一个dto比较好。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/controller/ActivityController.java:
```diff
@@ -0,0 +1,544 @@
+    }
+    permissionService.checkActivityOperationValidity(
+        commentPO.getActivityId(), idInToken, OperationTypeEnum.WRITE);
+    if (!commentService.updateById(new CommentPO(comment, idInToken))) {
```
注意检查隐藏的评论不能再被隐藏，感觉这种应该同时增加数据库的限制。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/controller/ActivityController.java:
```diff
@@ -0,0 +1,544 @@
+        description = "Add label to activity failed",
+        content = @Content(schema = @Schema(implementation = ErrorVO.class)))
+  })
+  public void addActivityLabel(
```
最好改成addlabeltoactivity。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/controller/ActivityController.java:
```diff
@@ -0,0 +1,544 @@
+    permissionService.checkActivityOperationValidity(
+        activityId, idInToken, OperationTypeEnum.WRITE);
+    // 检查这个标签是否存在于当前活动对应的仓库中
+    var activityPO = activityService.getById(activityId);
```
这个注释的位置没对，这个函数是不是也没有检查权限。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/controller/ActivityController.java:
```diff
@@ -0,0 +1,544 @@
+  public void addActivityLabel(
+      @RequestParam("activityId") Long activityId,
+      @RequestParam("labelId") Long labelId,
+      @RequestHeader(HeaderParameter.ACCESS_TOKEN) String accessToken) {
```
其实检查的逻辑可以按照从上到下的顺序来，例如查询活动的评论：

* 检查仓库
* 检查活动
* 检查评论

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/controller/ActivityController.java:
```diff
@@ -0,0 +1,544 @@
+          ErrorCodeEnum.ACTIVITY_NOT_FOUND, activityAssignLabelPO.getActivityId());
+    }
+    permissionService.checkRepositoryOperationValidity(
+        activityPO.getRepositoryId(), idInToken, OperationTypeEnum.WRITE);
```
同样的问题。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/controller/ActivityController.java:
```diff
@@ -0,0 +1,544 @@
+        description = "Add assignee to activity failed",
+        content = @Content(schema = @Schema(implementation = ErrorVO.class)))
+  })
+  public void addActivityAssignee(
```
检查活动是否存在。

luckygalaxy666 replied:
实际上检查活动是否存在是在permissionService.checkActivityOperationValidity中完成的，之所以在`addActivityLabel`中多了一段如下代码，是要通过activityPO来获取`repositoryId`,并通过`repositoryId`检查仓库中是否存在对应`labelId`的标签，而非用来检验活动是否存在的。
![image](https://github.com/user-attachments/assets/f5ab1ba4-8032-47ac-80f8-ebf33fee964e)
不过在之后的提交中，会要求前端提供repositoryId,所以`addActivityLabel`就不再需要通过activityPO来获取`repositoryId`了。

luckygalaxy666 committed:
fix(activity): fix some problems mentioned in comments

In this commit:
* Renamed parent_id to reply_to_id in t_acitivity_comment table.
* Separated the update activity locked and closed states from update
  activity and added the corresponding tests
* Separated the update comment resolved and hidden states from update
  comment and added ehe corresponding tests
* Realized functionality of sub-issue module
* Fixed the wrong logic of creating comment, now only repository creator
  and collaborators can creator comment when activity is locked
* Converted all properties in acitivtyDetailDTO to DTO form and add some
  constructors to convert DTO to VO.

In this next commit:
* move updateWrapper to service and set AOP point to remove redis cache
* modify the naming of some methods

luckygalaxy666 committed:
perf(activity): moved the use of updateWrapper into Service layer

In this commit,we moved the use of updateWrapper into Service layer
and add AOP method to delete cache for methods of removeBy* but not removeById.

luckygalaxy666 committed:
refactor(activity): refactored some codes mentioned in comments

In this commit:
* Renamed some methods in activityController and ActivityControllerTest.
* Renamed ActivityDetailDTO and ActivityDetailVO to ActivityFullInfoDTO
  and ActivityFullInfoVO.
* Added the creation time to the parameters of the paging query.
* Added validation messages of Activity Pojo.

luckygalaxy666 commented on src/main/java/edu/cmipt/gcs/controller/ActivityController.java:
```diff
@@ -0,0 +1,544 @@
+  public void createActivity(
+      @Validated(CreateGroup.class) @RequestBody ActivityDTO activity,
+      @RequestHeader(HeaderParameter.ACCESS_TOKEN) String accessToken)
+      throws InterruptedException {
```
sleep()方法要求抛interrupetdexception

luckygalaxy666 commented on src/main/java/edu/cmipt/gcs/service/ActivityServiceImpl.java:
```diff
@@ -0,0 +1,216 @@
+  @Transactional
+  public boolean removeById(Serializable id) {
+    // 删除活动时，删除活动相关的信息（标签、指定的参与者、评论等）
+    activityAssignLabelService.removeByActivityId(id);
```
目前所有removeByXXXId类型的方法的返回值都为List<Long>，表示对应数据库的主键id列表，通过`CacheAspect`实现了一个AOP方法通过获取方法返回值来删除对应key值的缓存。

luckygalaxy666 committed:
format(all): apply google java format automatically

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/util/TypeConversionUtil.java:
```diff
@@ -0,0 +1,30 @@
+  }
+
+  public static Long convertToLong(String value) {
+    return convertToLong(value, false, null);
```
删除, null并将此方法移到第一个。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/service/ActivityAssignLabelService.java:
```diff
@@ -0,0 +1,25 @@
+  Page<ActivityAssignLabelDTO> pageActivityLabelsByActivityId(
+      Long activityId, Page<ActivityAssignLabelDTO> objectPage);
+
+  ActivityAssignLabelPO getOneByActivityIdAndLabelId(Long activityId, Long labelId);
```
你还没删，点解决干嘛。

luckygalaxy666 committed:
fix(database): used partial unique index

In this commit,we fixed the bug mentioned in #167 by using partial
unique index.

luckygalaxy666 committed:
fix(all): fixed some problems mentioned in comments

In this commit:
* Refactored CacheAspect code.
* Added checkLabelNameValidity API and check the length of label name
  and description.
* Used try-catch to deal with duplicateKeyException when added duplicate
  asssignee or label to an activity.
* Removed some useless things.

luckygalaxy666 committed:
fix(activity): fix some minor issues found during self-reviews

In this commit:
* renamed validation messages param of checkNameValidity.
* renamed some unique index.
* renamed validationConAstant of check comment content length.
* deleted parameter in ActivityDTO: number, to avoid modify this in
  update method.
* replaced the id in the example with the id from the snowflake algorithm.
* fix some error description.

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/aop/CacheAspect.java:
```diff
@@ -105,21 +106,34 @@ public Object getOneAdvice(ProceedingJoinPoint joinPoint) throws Throwable {
       returning = "result")
   public void updateOrRemoveAdvice(JoinPoint joinPoint, Object result) throws Throwable {
-    if ((boolean) result) {
+    if (result instanceof Boolean) {
```
boolean instanceof Boolean是true吗？确认一下。

luckygalaxy666 replied:
`boolean instanceof Boolean` 这个语句在java中会报错，因为instanceof 不支持基元类型。
在AOP中，即使被拦截的方法返回 boolean（原始类型），Spring AOP 在内部也会将其自动装箱为 Boolean 对象（因为 AOP 的底层实现基于反射和动态代理，它们只能处理对象，不能直接处理原始类型。
我拿`updateActivityContent`作为例子，可以看到实际上`result`的类型是Boolean.
<img width="1263" height="138" alt="image" src="https://github.com/user-attachments/assets/2fe4bfd4-c1e9-4180-91cc-d94a855146e5" />
<img width="1957" height="76" alt="image" src="https://github.com/user-attachments/assets/5a2e1de6-07b9-465e-87ec-91e6e2be5637" />

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/aop/CacheAspect.java:
```diff
@@ -105,21 +106,34 @@ public Object getOneAdvice(ProceedingJoinPoint joinPoint) throws Throwable {
-              + "execution(* edu.cmipt.gcs.service.*ServiceImpl.removeById(java.io.Serializable))",
+          "execution(* edu.cmipt.gcs.service.*ServiceImpl.updateById(..)) || execution(*"
+              + " edu.cmipt.gcs.service.*ServiceImpl.removeBy*(..))"
+              + " ||execution(* edu.cmipt.gcs.service.*ServiceImpl.update*State(..))",
```
这里可以单独定义一个advice，因为你新加的这个不具有泛用性。

其他的类都没有 update*State 这个方法。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/constant/ApiPathConstant.java:
```diff
@@ -72,4 +81,46 @@ public class ApiPathConstant {
+      ACTIVITY_API_PREFIX + "/page-sub-issue";
+  public static final String ACTIVITY_CREATE_SUB_ISSUE_API_PATH =
+      ACTIVITY_API_PREFIX + "/create-sub-issue";
+  public static final String ACTIVITY_ADD_SUB_ISSUE_API_PATH =
```
这个是把一个已经有的问题改成另一个问题的子问题吗？如果是的话可以后面加一行注释，这个名字和上个接口的名字不好区分。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/constant/ValidationConstant.java:
```diff
@@ -16,6 +16,12 @@ public class ValidationConstant {
   public static final int MIN_REPOSITORY_DESCRIPTION_LENGTH = 0;
   public static final int MAX_REPOSITORY_DESCRIPTION_LENGTH = 255;
+
+  public static final int MIN_LABEL_NAME_LENGTH = 1;
```
注意检查常量的限制与数据库是否一致。

luckygalaxy666 committed:
fix(activity): fixed some problems mentioned in comments

In this commit:
* made constant limits are consistent with the database
* separated updateOneStateAdvice from updateOrRemoveAdvice
* renamed or fixed some variables.

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/aop/CacheAspect.java:
```diff
@@ -120,6 +121,27 @@ public void updateOrRemoveAdvice(JoinPoint joinPoint, Object result) throws Thro
+  }
+
+  @AfterReturning(
+      pointcut = "execution(* edu.cmipt.gcs.service.*ServiceImpl.update*State(..))",
```
这种的pointcut 就可以写的非常具体了，至少应该把类名的通配符去掉。

luckygalaxy666 committed:
fix(cache): redefined the pointcut of updateOnestateAdvice

luckygalaxy666 committed:
fix(activity): fix some bugs in the activity module

In this commit:
* Solved the problem that users can't create issue and comment in
other people's public repository.
* Added user full information to paged query comments for front-end
  presentation purposes.
* Added the API of checking activity operation validity for front-end
  presentation purposes.
* Added some tests.

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/controller/RepositoryController.java:
```diff
@@ -566,6 +560,146 @@ public PageVO<RepositoryVO> pageRepository(
+        description = "page labels failed",
+        content = @Content(schema = @Schema(implementation = ErrorVO.class)))
+  })
+  public PageVO<LabelVO> pageLabel(
```
分布查询应该限制深分布，可以看看以前怎么限制的。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/controller/RepositoryController.java:
```diff
@@ -566,6 +560,146 @@ public PageVO<RepositoryVO> pageRepository(
+    Long idInToken = TypeConversionUtil.convertToLong(JwtUtil.getId(accessToken), true);
+    permissionService.checkRepositoryOperationValidity(
+        repositoryId, idInToken, OperationTypeEnum.READ);
+    if (labelService.getOneByNameAndRepositoryId(labelName, repositoryId) != null) {
```
注意ignore case

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/enumeration/OperationTypeEnum.java:
```diff
@@ -0,0 +1,7 @@
+public enum OperationTypeEnum {
+  MODIFY,
+  READ,
+  CREATE // 包括创建活动和创建评论
```
改成 ATTACH 好一点。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/service/CommentService.java:
```diff
@@ -0,0 +1,27 @@
+  Page<CommentFullInfoDTO> pageCommentFullInfoByActivityId(
+      Integer page, Integer size, Long activityId);
+
+  Page<CommentFullInfoDTO> pageSubCommentFullInfoByActivityIdAndReplyToId(
```
是否直接用parent id就可以了。

Kaiser-Yang commented on src/main/java/edu/cmipt/gcs/pojo/comment/CommentFullInfoDTO.java:
```diff
@@ -0,0 +1,62 @@
+package edu.cmipt.gcs.pojo.comment;
```
这个类会和前端交互吗？

如果不应该不用写Schema

luckygalaxy666 committed:
fix(activity): fix some problems mentioned in comments

In this commit:
* Annotated all functions related to sub-problems because of insufficient
  consideration of permissions analysis.
* Renamed checkIssueOperationValidity to checkSubIssueCreationValidity
  and move this to ActivityController.
* Improved methods related to labels.
* refactors some codes

luckygalaxy666 committed:
fix(activity): fixed the permission control logic

In this commit,we fixed the permission related to activities and
comments and add some tests. You can consult the topmost comment of
the permissionServiceImpl for details.

Besides,I find a bug that if anyone updates the activity or comment,
it will change userId. It's wrong and we should add an external filed
called modifier_id and change this. I will fix it in this commit.

luckygalaxy666 committed:
fix(activity): fix code for update comment and activity

In this commit,we add modifier_id and rename user_id to creator_id
in activity.sql and activity_comment.sql and use this filed to record
modifier information instead of creator_id.Besides,add some variables
in CommentFullInfoVO and ActivityFullInfoVo so that we can jump to user
homepage in front-end.

luckygalaxy666 commented on src/main/java/edu/cmipt/gcs/pojo/activity/ActivityFullInfoVO.java:
```diff
@@ -0,0 +1,46 @@
+    @Schema(description = "Activity Description") String description,
+    @Schema(description = "Activity Labels") List<LabelVO> labels,
+    @Schema(description = "Creator Id") String creatorId,
+    @Schema(description = "Creator Username") String creatorUsername,
```
我就知道你要挑这种问题，我设计的前端不想展示用户详情信息，因为目前用户信息是非必要的，有名字作为展示，id作为跳转用户主页的条件足以。如果非要展示用户信息可以等后续用户详情信息设计后再展示，像github这样。
<img width="575" height="443" alt="image" src="https://github.com/user-attachments/assets/c1d94c79-43cc-4ba3-ae42-fd9bf57da4ba" />

luckygalaxy666 commented on src/main/java/edu/cmipt/gcs/pojo/activity/ActivityFullInfoVO.java:
```diff
@@ -0,0 +1,46 @@
+    @Schema(description = "Activity Labels") List<LabelVO> labels,
+    @Schema(description = "Creator Id") String creatorId,
+    @Schema(description = "Creator Username") String creatorUsername,
+    @Schema(description = "Modifier Id") String modifierId,
```
在前端中用于并且只支持展示最后一次修改者的信息的，所以时间就是GmtUpdated。没有url原因同上。

luckygalaxy666 commented on src/main/java/edu/cmipt/gcs/pojo/comment/CommentFullInfoVO.java:
```diff
@@ -0,0 +1,40 @@
+    @Schema(description = "Comment ID") String id,
+    @Schema(description = "Activity ID") String activityId,
+    @Schema(description = "Creator ID") String creatorId,
+    @Schema(description = "Creator Username") String creatorUsername,
```
同理

luckygalaxy666 commented on src/main/java/edu/cmipt/gcs/pojo/label/LabelVO.java:
```diff
@@ -0,0 +1,18 @@
+import io.swagger.v3.oas.annotations.media.Schema;
+
+@Schema(description = "Label Data Transfer Object")
+public record LabelVO(String id, String name, String hexColor, String description) {
```
目前不需要，道理同上

luckygalaxy666 committed:
feat(activity): added user information for front-end rendering

luckygalaxy666 committed:
format(all): apply google java format automatically
